src/utils/Process.ts

```diff
@@ -0,0 +1,102 @@
+import {executeCommandWithShellConfig} from "../PTY";
+import {memoize, intersection} from "lodash";
+
+// http://linuxcommand.org/man_pages/ps1.html
+//   ps -eo euser,ruser,euid,ruid,comm
+//   ps -eo rgid,egid,rgroup,egroup,comm
+
+export interface User {
+    ruserid: string;
+    ruser: string;
+    euserid: string;
+    euser: string;
+}
+
+export interface Group {
+    rgroupid: string;
+    rgroup: string;
+    egroupid: string;
+    egroup: string;
+}
+
+export interface Terminal {
+    name: string;
+}
+
+export interface Process {
+    pid: string;
+    time: string;
+    ruser: string;
+    cmd: string;
+}
+
+export interface Session {
+    sid: string;
+    rgroup: string;
+}
+
+const ignoreUsers: string[] = ["nobody"];
+
+const ignoreGroups: string[] = ["nobody"];
+
+export const users =
+    async(): Promise<User[]> => {
+        const pInfo: string[] = await executeCommandWithShellConfig("ps -eo ruid,ruser,euid,euser");
+        return pInfo.splice(1)
+                    .map(p => {
+                        const v = p.trim().replace(/ +(?= )/g, "").split(" ");
+                        return <User> {ruserid: v[0], ruser: v[1], euserid: v[2], euser: v[3]};
+                    })
+                    .filter(p => intersection(ignoreUsers, [p.ruser, p.euser]).length === 0);
+    };
+
+export const groups =
+    async(): Promise<Group[]> => {
+    const pInfo: string[] = await executeCommandWithShellConfig("ps -eo rgid,rgroup,egid,egroup");
+    return pInfo.splice(1)
+                .map(p => {
+                    const v = p.trim().replace(/ +(?= )/g, "").split(" ");
+                    return <Group> {rgroupid: v[0], rgroup: v[1], egroupid: v[2], egroup: v[3]};
+                })
+                .filter(p => intersection(ignoreGroups, [p.rgroup, p.egroup]).length === 0);
+    };
+
+export const terminals =
+    async(): Promise<Terminal[]> => {
+        const pInfo: string[] = await executeCommandWithShellConfig("ps -eo tty");
+        return pInfo.splice(1)
+                    .filter(p => p.trim() !== "?")
+                    .map(p => <Terminal> {name: p.trim()});
+    };
+
+export const processes =
+    async(): Promise<Process[]> => {
+        const pInfo: string[] = await executeCommandWithShellConfig("ps -eo pid,time,ruser,cmd");
+        return pInfo.splice(1)
+                    .map(p => {
+                        const v = p.trim().replace(/ +(?= )/g, "").split(" ", 4);
+                        return <Process> {pid: v[0], time: v[1], ruser: v[2], cmd: v[3]};
+                    });
+    };
+/*
```

Comment:
Attempted to cache results of "processes" function for each command instance. Unfortunately this caches results for entire terminal session :(